pkg/go/graph/weighted_graph.go

```diff
@@ -168,6 +183,45 @@ func (wg *WeightedAuthorizationModelGraph) isLogicalUnionOperator(node *Weighted
 	return (nodeType == OperatorNode && node.GetLabel() == UnionOperator) || nodeType == LogicalTTUGrouping || nodeType == LogicalDirectGrouping
 }
 
+// GetWeightForUserset returns the weight for the given userset node.
+func (wg *WeightedAuthorizationModelGraph) getWeightForUserset(node *WeightedAuthorizationModelNode, userset string) (int, bool) {
+	// Check for cached result first
+	if len(node.usersetWeights) != 0 {
+		weight, exists := node.usersetWeights[userset]
+		if exists {
+			if weight > 0 {
+				return weight, true
+			}
+			return 0, false
+		}
+	}
+
+	// Get the userset node
+	usersetNode := wg.nodes[userset]
+	if usersetNode == nil || len(node.weights) == 0 || len(usersetNode.weights) == 0 {
+		wg.setUsersetWeightToNode(node, userset, 0)
+		return 0, false
+	}
+
+	// Early pruning: Check if the node can reach the userset based on terminal type weights
+	// This avoids unnecessary traversal if we can determine that no path exists
+	for key, usersetValue := range usersetNode.weights {
+		nodeValue, exists := node.weights[key]
+		if !exists || nodeValue < usersetValue || (nodeValue == usersetValue && nodeValue != Infinite) {
```

Comment:
Can you call out what does `nodeValue < usersetValue` and (nodeValue == usersetValue && nodeValue != Infinite) mean? Why do we need to set the node's userset value to 0 for these cases?